src/librbd/cache/rwl/InitRequest.cc

```diff
@@ -0,0 +1,217 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/cache/rwl/InitRequest.h"
+#include "librbd/Utils.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/asio/ContextWQ.h"
+#include "librbd/cache/rwl/ImageCacheState.h"
+#include "librbd/cache/ReplicatedWriteLog.h"
+
+#define dout_subsys ceph_subsys_rbd_rwl
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::rwl:InitRequest " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace cache {
+namespace rwl {
+
+using util::create_async_context_callback;
+using util::create_context_callback;
+
+template <typename I>
+InitRequest<I>* InitRequest<I>::create(I &image_ctx,
+                                       Context *on_finish) {
+  return new InitRequest(image_ctx, on_finish);
+}
+
+template <typename I>
+InitRequest<I>::InitRequest(I &image_ctx, Context *on_finish)
+  : m_image_ctx(image_ctx),
+    m_on_finish(create_async_context_callback(image_ctx, on_finish)),
+    m_error_result(0) {
+}
+
+template <typename I>
+InitRequest<I>::~InitRequest() {
+}
+
+template <typename I>
+bool InitRequest<I>::is_rwl_enabled() {
+  return m_image_ctx.config.template get_val<bool>("rbd_rwl_enabled");
+}
+
+template <typename I>
+void InitRequest<I>::send() {
+  send_pre_check();
+}
+
+
+template <typename I>
+void InitRequest<I>::send_pre_check() {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  bool is_dirty_cache = false;
+  if (m_image_ctx.test_features(RBD_FEATURE_IMAGE_CACHE)) {
+    is_dirty_cache = true;
+  }
+  bool rwl_enabled = is_rwl_enabled();
+  ldout(cct, 10) << "rwl_enabled=" << rwl_enabled << dendl;
+  if(!is_dirty_cache && !rwl_enabled) {
+    finish();
+    return;
+  }
+  // in fact, we handle this case in get_image_cache_state()
+  if(is_dirty_cache && !rwl_enabled) {
+    lderr(cct) << "There's a dirty cache, but RWL cache is disabled in ceph.conf"
+               << dendl;
+    save_result(-EINVAL);
+    finish();
+    return;
+  }
+
+  Context *ctx = create_context_callback<
+    InitRequest<I>, &InitRequest<I>::handle_pre_check>(this);
+  ctx->complete(0);
+}
+
+template <typename I>
+void InitRequest<I>::handle_pre_check(int r) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << ": r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to pre_check" << cpp_strerror(r)
```

Comment:
Nit: no spacing between text and error translation

Reply:
The function was removed.